sub/filter.c

```diff
@@ -0,0 +1,420 @@
+{
+    unsigned char *end = *endp;
+
+    skip_ass(sd, &wp, end);
+    if (wp < end && *wp != '-')
+        return;
+    unsigned char *rp = wp + 1;
+    copy_ass(sd, &rp, end, &wp);
+    while (rp < end && *rp == ' ') {
+        rp++;
+        copy_ass(sd, &rp, end, &wp);
+    }
+    while (rp < end) {
+        *wp++ = *rp++;
+    }
+    *endp = wp;
+}
+
+// Filter ASS formatted string for SDH
+//
+// Parameters:
+//     format       format line from ASS configuration
+//     n_ignored    number of comma to skip as preprocessing have removed them
+//     ass          ASS line.
+//     length       length of ASS string in bytes
+//
+// Returns  a talloc allocated string with filtered ASS data (may be the same
+// as original if no SDH was found) which must be released by caller using
+// talloc_free.
+//
+// Returns NULL if filtering resulted in all of ASS data being removed so no
+// subtitle should be output
+char *filter_SDH(struct sd *sd, char *format, int n_ignored, char *ass, int length)
+{
+    unsigned char *rp = ass;
+    unsigned char *end = ass + length;
+
+    if (!format) {
+        MP_WARN(sd, "SDH filtering not possible - format missing\n");
+        return talloc_strndup(NULL, ass, length);
+    }
+
+    unsigned char *buffer = talloc_size(NULL, length + 1); // with room for terminating '\0'
+    unsigned char *wp = buffer;
+
+    int comma = 0;
+    // scan format line to find the number of the field where the text is
+    for (char *c = format; *c; c++) {
+        if (*c == ',') {
+            comma++;
+            if (strncasecmp(c + 1, "Text", 4) == 0)
+                break;
+        }
+    }
+    // if preprocessed line some fields are skipped
+    comma -= n_ignored;
+
+    // locate text field in ASS line
+    for (int k = 0; k < comma; k++) {
+        while (rp < end && *rp != ',')
+            *wp++ = *rp++;
+        if (rp < end)
+            *wp++ = *rp++;
+    }
+    if (rp == end) {
+        talloc_free(buffer);
+        MP_WARN(sd, "SDH filtering not possible - cannot find text field\n");
+        return talloc_strndup(NULL, ass, length);
+    }
+
+    bool contains_text = false;  // true if non SDH text was found
+    bool line_with_text = false; // if last line contained text
+    unsigned char *wp_line_start = wp;    // write pointer to start of last line
+    unsigned char *wp_line_end   = wp;    // write pointer to end of previous line with text (\N)
+    unsigned char *start_wp = wp;         // write pointer to start of text place
+
+    // go through the lines in the text
+    // they are separated by \N
+    while (rp < end) {
+        line_with_text = false;
+        wp_line_start = wp;
+
+        // skip any speaker label
+        skip_speaker_label(sd, &rp, end, &wp);
+
+        // go through the rest of the line looking for SDH in () or []
+        while (rp < end && !(rp < end-1 && *rp == '\\' && rp[1] == 'N')) {
```

Comment:
Yes, I can do that. Though I see that other files in mpv does not do that. For example in osd_libass.c